Fix unit tests and make tests fail if a promise rejects after the test harness is finishes

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit tests and make tests fail if a promise rejects after the test harness is finishes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 commented on package.json:
```diff
@@ -25,7 +25,7 @@
     "build:mac": "yarn run compile && electron-builder --mac --dir -c.productName=Lens",
     "build:win": "yarn run compile && electron-builder --win --dir -c.productName=Lens",
-    "test": "jest --env=jsdom src $@",
+    "test": "scripts/test.sh",
```
This helps support `yarn test <file-name>` only testing one file

jakolehm replied:
I don't really like that we split these into yet-another script (worst case: make -> yarn -> bash). How about:

```
    "jest-jsdom": "jest --env=jsdom",
    "test": "yarn run jest-jsdom src",
```

Now you could run custom targets using `yarn jest-jsdom path/to/file.test.ts`.

aleksfront replied:
Probably we should't do such changes. If we eliminate `"testEnvironment": "node",` from our `package.json`, tests still works okay and `jsdom` set as default value. So we don't need passing extra vars `--env=jsdom` or
```
/**
 * @jest-environment jsdom
 */
```
declaration in specific files.

Nokel81 replied:
Ah so we could just remove the "test" script all together then, brilliant!

Nokel81 commented on src/renderer/api/__tests__/kube-api-parse.test.ts:
```diff
@@ -1,134 +1,115 @@
-      expect(parseKubeApi(url)).toStrictEqual(expected);
-    });
-  }
+  it.each(tests)("testing %s", (url, expected) => {
+    expect(parseKubeApi(url)).toStrictEqual(expected);
+  });
```
This just makes use of the jest table testing format, these specific tests weren't causing the problems.

jakolehm commented on src/common/base-store.ts:
```diff
@@ -19,7 +19,7 @@ export interface BaseStoreParams<T = any> extends ConfOptions<T> {
  */
 export abstract class BaseStore<T = any> extends Singleton {
-  protected storeConfig: Config<T>;
+  protected storeConfig?: Config<T>;
```
Why are we modifying `BaseStore`?

Nokel81 replied:
Because one of the errors I encountered was due to a call to `onModelChange` in the `user-store.test.ts` happening after the test suite was finished (because we don't handle rejected promises very well) and since in that test `isLoaded` was never true the function `enableSync` wasn't called so `storeConfig` was never set.

These changes merely make `BaseStore` more robust by marking the types that get late initialized as such.

jakolehm commented on src/renderer/components/dock/__test__/dock-tabs.test.tsx:
```diff
@@ -8,6 +8,20 @@ import { createResourceTab } from "../create-resource.store";
 import { createTerminalTab } from "../terminal.store";
 import { observable } from "mobx";
 
+Object.defineProperty(window, "matchMedia", {
```
This is the actual fix for unhandled exceptions, right?

Nokel81 replied:
To some of them. I was also getting "KubeApi.parseApi is not a function" which is why there are those mocks in "kube-api-parse.test.ts".

aleksfront replied:
This test is fixed without needed to mock `matchMedia`.
https://github.com/lensapp/lens/pull/2037. But we can remeber such cases where cleaning test file isn't enough.

aleksfront replied:
Again, this should be removed for now. Test passes without these changes ^

Nokel81 replied:
Done.

aleksfront commented on scripts/test.sh:
```diff
@@ -0,0 +1 @@
+jest --env=jsdom ${1:-src}
```
No need for this file (see my previous comment).

Nokel81 replied:
Sounds good

aleksfront reviewed (changes requested):
Most of the errors are gone which is really great! But I've got another one introduced (related to xterm somehow). They're appeared after few test files (not only one presented in screenshot).

<img width="1217" alt="xterm errors" src="https://user-images.githubusercontent.com/9607060/106148054-a618f780-6189-11eb-8439-d3c6e7fdf099.png">

@Nokel81 Can you please run tests again and check this one?

Nokel81 committed:
General changes

- Remove scripts/test.sh and move functionality into `Makefile`

- improve rebuilting ability of Makefile with better dependencies

- change default jest env to jsdom, remove all file headers except for
  integration test file which should be run with jest env node

- Remove package.json test script as `make test` covers that

Signed-off-by: Sebastian Malton <sebastian@malton.name>

aleksfront commented on src/common/__tests__/cluster-store.test.ts:
```diff
@@ -35,7 +35,8 @@ jest.mock("electron", () => {
       getPath: () => "tmp",
-      getLocale: () => "en"
+      getLocale: () => "en",
+      setLoginItemSettings: jest.fn(),
```
Can you please remember why this line needed? I've checked test with/without it, and results are equal.

Nokel81 replied:
This is required if a `mobx` reaction fires to a change on the userPreferences.openAtLogin. Since that reaction is classified as "fireImmediately" and is in an async function, it could fire at any time.

aleksfront reviewed (commented):
Posted minor cleaning comments.

@Nokel81 Changes introduced in `Makefile` isn't unclear to me. Can you please explain why we added `node_modules` part to most of the commands, and why we need to add `CMD_ARGS`?

Nokel81 committed:
remove mocked matchMedia

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on Makefile:
```diff
@@ -9,23 +14,24 @@ else
     DETECTED_OS := $(shell uname)
 endif
 
+.PHONY: binaries/client
```
Why do we want to download client binaries on every run?

Nokel81 replied:
I think I added this because it doesn't doesn't check correctly the folder modification date. Plus the downloaders already check and verify the current versions before downloading again.

Nokel81 committed:
remove PHONY

Signed-off-by: Sebastian Malton <sebastian@malton.name>
